crates/typst-library/src/diag.rs

```diff
@@ -568,31 +571,270 @@ impl From<PackageError> for EcoString {
     }
 }
 
+pub type LoadResult<T> = Result<T, LoadError>;
+
+/// A call site independent error that occurred during data loading.
+/// This avoids polluting the memoization with [`Span`]s and [`FileId`]s from source files.
+/// Can be turned into a [`SourceDiagnostic`] using the [`LoadedAt::in_text`]
+/// or [`LoadedAt::in_invalid_text`] methods available on [`LoadResult`].
+///
+/// [`FileId`]: typst_syntax::FileId
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct LoadError {
+    pub pos: ReportPos,
+    pub message: EcoString,
+    pub error: EcoString,
+}
+
+impl LoadError {
+    pub fn new(
+        pos: impl Into<ReportPos>,
+        message: impl std::fmt::Display,
+        error: impl std::fmt::Display,
+    ) -> Self {
+        Self {
+            pos: pos.into(),
+            message: eco_format!("{message}"),
+            error: eco_format!("{error}"),
+        }
+    }
+}
+
+/// Convert a [`LoadResult`] to a [`SourceResult`] by adding the [`Loaded`] context.
+pub trait LoadedAt<T> {
+    /// Add the span information.
+    fn in_text(self, data: &Loaded) -> SourceResult<T>;
+
+    /// Add the span information.
+    fn in_invalid_text(self, data: &Loaded) -> SourceResult<T>;
```

Comment:
It wasn't immediately clear to me when to call what as a normal user of this infrastructure. It seems like `in_invalid_text` is only used in `load_str`. Maybe the trait should be renamed `LoadedWithin` and only expose a single `.within` function that just does the right thing?

in particular, the code still works if I replace the usage of `in_invalid_text` with `in_text`, so it's just a small perf optimization, right? I would probably 
- either opt for simpler single-function API instead 
- or at least not expose the optimized variant in this trait in a way that forces consumers of the API to make a decision 
- or, as a final option, document when to use what. :)